grid/src/Cabana_Grid_Splines.hpp

```diff
@@ -451,6 +451,586 @@ struct Spline<3>
     }
 };
 
+//! Quartic. Defined on the dual grid.
+template <>
+struct Spline<4>
+{
+    //! Order.
+    static constexpr int order = 4;
+
+    //! The number of non-zero knots in the spline.
+    static constexpr int num_knot = 5;
+
+    /*!
+      \brief Map a physical location to the logical space of the primal
+      grid in a single dimension.
+      \param xp The coordinate to map to the logical space.
+      \param rdx The inverse of the physical distance between grid locations.
+      \param low_x The physical location of the low corner of the primal grid.
+      \return The coordinate in the logical primal grid space.
+
+      \note Casting this result to an integer yields the index at the center of
+      the stencil.
+      \note A quartic spline uses the dual grid.
+    */
+    template <class Scalar>
+    KOKKOS_INLINE_FUNCTION static Scalar
+    mapToLogicalGrid( const Scalar xp, const Scalar rdx, const Scalar low_x )
+    {
+        return ( xp - low_x ) * rdx + 0.5;
+    }
+
+    /*!
+     *       \brief Get the logical space stencil offsets of the spline. The
+     * stencil defines the offsets into a grid field about a logical coordinate.
+     *                   \param indices The stencil index offsets.
+     *                       */
+    KOKKOS_INLINE_FUNCTION
+    static void offsets( int indices[num_knot] )
+    {
+        indices[0] = -2;
+        indices[1] = -1;
+        indices[2] = 0;
+        indices[3] = 1;
+        indices[4] = 2;
+    }
+
+    /*!
+      \brief Compute the stencil indices for a given logical space location.
+      \param x0 The coordinate at which to evaluate the spline stencil.
+      \param indices The indices of the stencil.
+    */
+    template <class Scalar>
+    KOKKOS_INLINE_FUNCTION static void stencil( const Scalar x0,
+                                                int indices[num_knot] )
+    {
+        indices[0] = static_cast<int>( x0 ) - 2;
+        indices[1] = indices[0] + 1;
+        indices[2] = indices[1] + 1;
+        indices[3] = indices[2] + 1;
+        indices[4] = indices[3] + 1;
+    }
+
+    /*!
+      \brief Calculate the value of the spline at all knots.
+      \param x0 The coordinate at which to evaluate the spline in the logical
+      grid space.
+      \param values Basis values at the knots. Ordered from lowest to highest
+      in terms of knot location.
+    */
+    template <typename Scalar>
+    KOKKOS_INLINE_FUNCTION static void value( const Scalar x0,
+                                              Scalar values[num_knot] )
+    {
+        // Constants
+        Scalar denom_2 = 1.0 / 384.0;
+        Scalar denom_1 = denom_2 * 4.0;
+        Scalar denom_0 = denom_2 * 2.0;
+
+        // Knot at i - 2
+        Scalar xn = x0 - static_cast<int>( x0 );
+        xn -= 0.5;
+        values[0] =
```

Comment:
Clean up the formatting on this so it is easier to read. Using the pre-calcuated powers of xn as mentioned above should help